walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -14,10 +14,185 @@
+    if (request.getDuration() > TimeRange.WHOLE_DAY.duration()){
+      return Arrays.asList();
+    }
+    // No events/no attendess means meeting can be anytime.
+    if(events.isEmpty() || (mandatoryBookedEvents.isEmpty() && optionalBookedEvents.isEmpty())) {
+        return Arrays.asList(TimeRange.WHOLE_DAY);
+    } 
+
+    // If mandatory attendees are available all day with no booked events.
+    // Try and find available times including optional event restrictions and return unless.
+    // no time that works with optional attendess works just return mandatory availabilities (All day).
+    if(mandatoryBookedEvents.isEmpty()){
+      Collection<TimeRange> allAvailable = findPotentialTimes(optionalBookedEvents, request);
+      
+      if(allAvailable.isEmpty()){
+        return Arrays.asList(TimeRange.WHOLE_DAY);
+      }
+      return allAvailable;
+    }
+
+    // Try finding time with all attendees, if none can be found try finding time with
+    // only mandatory attendees.
+    Collection<TimeRange> times = findPotentialTimes(allBookedEvents, request);
+    if(times.isEmpty()){
+      return findPotentialTimes(mandatoryBookedEvents, request);
+    }
+    return times;
   }
+
+  /**
+   * Finds all the events that attendees in the request are booked for.
+   * @param events - all possible events.
+   * @param attendees - from the meeting request.
+   * @return events that request attendees are booked for.
+   */
+  public List<Event> findBookedEvents(Collection<Event> events, Collection<String> attendees) {
+    // List of existing bookedEvents for attendees in meeting request.
+    List<Event> bookedEvents = new ArrayList<>();
+
+    // Find all events that attendees in the request are already booked for.
+    for(Event event : events){
+      for (String attendee : attendees){
+        if(event.getAttendees().contains(attendee) && !bookedEvents.contains(event)){
+          bookedEvents.add(event);
+        }
+      }
+    }
+    return bookedEvents;
+  } 
+  /**
+   * Finds all potential availabilities in between bookedEvents.
+   * @param bookedEvents - timerange already booked out/unavailable.
+   * @param request - meetin request to add in a potential new timerange.
+   * @return list of timeranges for availability.
+   */
+  public Collection<TimeRange> findPotentialTimes(List<Event> bookedEvents, MeetingRequest request){
+    // List of potential free timeranges for the meeting request.
+    List<TimeRange> potentialTime = new ArrayList<>();
+    
+    Event prevEvent = bookedEvents.get(0); 
+    Event currEvent = bookedEvents.get(0);
+    for(int i = 0; i < bookedEvents.size(); i++){ 
+      prevEvent = currEvent; 
+      currEvent = bookedEvents.get(i); 
+
+      // If only one event - consider only timerange before and after the event.
+      if(bookedEvents.size() == 1) {
+        potentialTime.add(TimeRange.fromStartEnd(TimeRange.START_OF_DAY, currEvent.getWhen().start(), false));
```

Comment:
what if there's not enough time for the meeting between the start or the end of the day?

Reply:
Done.